lib/bookmark_manager.rb

```diff
@@ -7,22 +7,27 @@ def initialize(user)
     @user = user
   end
 
-  def create(post_id:, name: nil, reminder_type: nil, reminder_at: nil, options: {})
-    post = Post.find_by(id: post_id)
+  def create(post_id:, topic_id: nil, name: nil, reminder_type: nil, reminder_at: nil, options: {})
+    if post_id.blank? && topic_id.present?
+      bookmark_for_topic = true
+      topic = Topic.find_by(id: topic_id)
+    else
+      post = Post.find_by(id: post_id)
+      topic = post&.topic
+    end
     reminder_type = parse_reminder_type(reminder_type)
 
-    # no bookmarking deleted posts or topics
-    raise Discourse::InvalidAccess if post.blank? || post.topic.blank?
+    raise Discourse::InvalidAccess if (!bookmark_for_topic && post.blank?) || topic.blank?
 
-    if !Guardian.new(@user).can_see_post?(post) || !Guardian.new(@user).can_see_topic?(post.topic)
+    if (!bookmark_for_topic && !Guardian.new(@user).can_see_post?(post)) || !Guardian.new(@user).can_see_topic?(topic)
       raise Discourse::InvalidAccess
     end
 
     bookmark = Bookmark.create(
       {
         user_id: @user.id,
-        topic: post.topic,
-        post: post,
+        topic: topic,
```

Comment:
Technically if `topic_id` is nil and we're using `post_id` to fetch the topic, we need the fetching and creating to be in the same transaction.